src/stan/mcmc/chainset.hpp

```diff
@@ -0,0 +1,556 @@
+  std::vector<Eigen::MatrixXd> chains_;
+
+  static size_t thinned_samples(const stan::io::stan_csv& stan_csv) {
+    size_t thinned_samples = stan_csv.metadata.num_samples;
+    if (stan_csv.metadata.thin > 0) {
+      thinned_samples = thinned_samples / stan_csv.metadata.thin;
+    }
+    return thinned_samples;
+  }
+
+  static bool is_valid(const stan::io::stan_csv& stan_csv) {
+    if (stan_csv.header.empty())
+      return false;
+    if (stan_csv.samples.size() == 0)
+      return false;
+    if (stan_csv.samples.rows() != thinned_samples(stan_csv))
+      return false;
+    return true;
+  }
+
+  /**
+   * Process first chain: record header, thinned samples,
+   * add samples to vector chains.
+   */
+  void init_from_stan_csv(const stan::io::stan_csv& stan_csv) {
+    if (!is_valid(stan_csv)) {
+      throw std::invalid_argument("Invalid sample");
+    }
+    if (chains_.size() > 0) {
+      throw std::invalid_argument("Cannot re-initialize chains object");
+    }
+    param_names_ = stan_csv.header;
+    num_samples_ = thinned_samples(stan_csv);
+    chains_.push_back(stan_csv.samples);
+  }
+
+  /**
+   * Process next chain: validate size, shape, column names,
+   * append to vector chains.
+   */
+  void add(const stan::io::stan_csv& stan_csv) {
+    if (!is_valid(stan_csv)) {
+      throw std::invalid_argument("Invalid sample");
+    }
+    if (stan_csv.header.size() != num_params()) {
+      throw std::invalid_argument(
+          "Error add(stan_csv): number of columns in"
+          " sample does not match first chain");
+    }
+    if (thinned_samples(stan_csv) != num_samples_) {
+      throw std::invalid_argument(
+          "Error add(stan_csv): number of sampling draws in"
+          " sample does not match first chain");
+    }
+    for (int i = 0; i < num_params(); i++) {
+      if (param_names_[i] != stan_csv.header[i]) {
+        std::stringstream ss;
+        ss << "Error add(stan_csv): header " << param_names_[i]
+           << " does not match chain's header (" << stan_csv.header[i] << ")";
+        throw std::invalid_argument(ss.str());
+      }
+    }
+    chains_.push_back(stan_csv.samples);
+  }
+
+ public:
+  explicit chainset(const stan::io::stan_csv& stan_csv) {
+    init_from_stan_csv(stan_csv);
+  }
+
+  explicit chainset(const std::vector<stan::io::stan_csv>& stan_csv) {
+    if (stan_csv.empty())
+      return;
+    init_from_stan_csv(stan_csv[0]);
+    for (size_t i = 1; i < stan_csv.size(); ++i) {
+      add(stan_csv[i]);
+    }
+  }
+
+  /**
+   * Report number of chains in chainset.
+   * @return chainset size.
+   */
+  inline int num_chains() const { return chains_.size(); }
+
+  /**
+   * Report number of parameters per chain.
+   * @return size of parameter names vector.
+   */
+  inline int num_params() const { return param_names_.size(); }
+
+  /**
+   * Report number of samples (draws) per chain.
+   * @return rows per chain
+   */
+  inline int num_samples() const { return num_samples_; }
```

Comment:
`std::vector` returns `size_t` which is a `unsigned int`. I like using `Eigen::Index` instead of `int` since it's a `long int` and it's a reasonably safe int to convert to

```suggestion
  inline int num_chains() const { return chains_.size(); }

  /**
   * Report number of parameters per chain.
   * @return size of parameter names vector.
   */
  inline int num_params() const { return param_names_.size(); }

  /**
   * Report number of samples (draws) per chain.
   * @return rows per chain
   */
  inline int num_samples() const { return num_samples_; }
```